sign/dilithium/internal/ntt.go

```diff
@@ -0,0 +1,190 @@
+	1851402, 5220671, 3553272, 8190869, 1316856, 7620448, 210977,
+	5991061, 3249728, 6727353, 8578, 3724342, 4421799, 7475901,
+	1100098, 8336129, 5282425, 7871466, 8115473, 3343383, 1430430,
+	6527646, 7031341, 381987, 1308169, 22981, 1228525, 671102,
+	2477047, 411027, 3693493, 2967645, 5665122, 6232521, 983419,
+	4968207, 8253495, 3632928, 3157330, 3190144, 1000202, 4083598,
+	6441103, 1257611, 1585221, 6203962, 4904467, 1452451, 3041255,
+	3677745, 1528703, 3930395, 2797779, 6308525, 2556880, 4479693,
+	4499374, 7426187, 7849063, 7568473, 4680821, 1600420, 2140649,
+	4873154, 3821735, 4874723, 1643818, 1699267, 539299, 6031717,
+	300467, 4840449, 2867647, 4805995, 3043716, 3861115, 4464978,
+	2537516, 3592148, 1661693, 4849980, 5303092, 8284641, 5674394,
+	8100412, 4369920, 19422, 6623180, 3277672, 1399561, 3859737,
+	2118186, 2108549, 5760665, 1119584, 549488, 4794489, 1079900,
+	7356305, 5654953, 5700314, 5268920, 2884855, 5260684, 2091905,
+	359251, 6026966, 6554070, 7913949, 876248, 777960, 8143293,
+	518909, 2608894, 8354570, 4186625,
+}
+
+// Execute an in-place forward NTT on as.
+//
+// Assumes the coefficients are in Montgomery representation and bounded
+// by 2*Q.  The resulting coefficients are again in Montgomery representation,
+// but are only bounded bt 18*Q.
+func (p *Poly) NTT() {
+	// Writing z := zeta for our root of unity zeta := 1753, note z^256=-1
+	// (otherwise the order of z wouldn't be 512) and so
+	//
+	//  x^256 + 1 = x^256 - z^256
+	//            = (x^128 - z^128)(x^128 + z^128)
+	//            = (x^64 - z^64)(x^64 + z^64)(x^64 + z^192)(x^64 - z^192)
+	//          ...
+	//            = (x-z)(x+z)(x - z^129)(x + z^129) ... (x - z^255)(x + z^255)
+	//
+	// Note that the powers of z that appear (from the second line) are
+	//  in binary
+	//
+	//  01000000 11000000
+	//  00100000 10100000 01100000 11100000
+	//  00010000 10010000 01010000 11010000 00110000 10110000 01110000 11110000
+	//     ...
+	//
+	// i.e. brv(2), brv(3), brv(4), ... and these powers of z are given by
+	// the Zetas array.
+	//
+	// The polynomials x ± z^i are irreducable and coprime, hence by the
+	// Chinese Remainder Theorem we know
+	//
+	//  R[x]/(x^256+1) ---> R[x] / (x-z) x ... x R[x] / (x+z^255)
+	//                      ~= ∏_i R
+	//
+	// given by
+	//
+	//  a |---> = ( a mod x-z, ..., a mod x+z^255 )
+	//          ~ ( a(z), a(-z), a(z^129), a(-z^129), ..., a(z^255), a(-z^255) )
+	//
+	// is an isomorphism, which is the forward NTT.  It can be computed
+	// efficiently by computing
+	//
+	//  a |---> ( a mod x^128 - z^128, a mod x^128 + z^128 )
+	//    |---> ( a mod x^64 - z^64, a mod x^64 + z^64,
+	//            a mod x^64 - z^192, a mod x^64 + z^192 )
+	//       et cetera
+
+	k := 0 // Index into Zetas
+
+	for l := uint(N / 2); l > 0; l >>= 1 {
+		// On the n-th iteration of the l-loop, the coefficients start off
+		// bounded by n*2*Q.
+		for offset := uint(0); offset < N-l; offset += 2 * l {
+			k++
```

Comment:
This picture illustrates what happens for a smaller NTT.

![File0026](https://user-images.githubusercontent.com/9975/77101950-04286a80-6a10-11ea-87be-1dc180df3bb1.png)

Each cross is a Cooley–Tukey butterfly: it's the map `(a,b) |-> (a + zeta b, a - zeta b)` for the appropriate zeta for that column and row group.

The outer loop `l` effectively loops over the columns.  `l` is half the height of a row group, i.e. the number of butterflies in each row group.  In this diagram it would be 4, 2, 1.

The middle loop `offset` effectively loops over the row groups.  `offset` is the first row of the row group.

The inner loop `j` loops over the butterflies in the row group.

Reply:
What do you think, should I expand my comment with an ASCII rendition of such a diagram?

Reply:
I've added the comments.

Reply:
@wbl is it clearer now?